src/main/java/com/rits/cloning/Cloner.java

```diff
@@ -417,97 +407,226 @@ private boolean isImmutable(final Class<?> clz) {
-				if (!(nullTransient && Modifier.isTransient(modifiers))) {
-					// request by Jonathan : transient fields can be null-ed
-					final Object fieldObject = field.get(o);
-					final boolean shouldClone = (cloneSynthetics || !field.isSynthetic()) && (cloneAnonymousParent || !isAnonymousParent(field));
-					final Object fieldObjectClone = clones != null ? (shouldClone ? applyCloningStrategy(clones, o, fieldObject, field) : fieldObject) : fieldObject;
-					field.set(newInstance, fieldObjectClone);
-					if (dumpCloned != null && fieldObjectClone != fieldObject) {
-						dumpCloned.cloning(field, o.getClass());
+
+		public <T> T deepClone(T o, Map<Object, Object> clones) {
+			if (dumpCloned != null) {
+				dumpCloned.startCloning(o.getClass());
+			}
+			int length = Array.getLength(o);
+			T newInstance = (T) Array.newInstance(componentType, length);
+			if (clones != null) {
+				clones.put(o, newInstance);
+			}
+			if (primitive || immutable) {
+				System.arraycopy(o, 0, newInstance, 0, length);
+			} else {
+				if (clones == null) {
+					for (int i = 0; i < length; i++) {
+						Array.set(newInstance, i, Array.get(o, i));
+					}
+				} else {
+					for (int i = 0; i < length; i++) {
+						Array.set(newInstance, i, cloneInternal(Array.get(o, i), clones));
 					}
 				}
 			}
+			return newInstance;
 		}
-		return newInstance;
 	}
 
-	private Object applyCloningStrategy(Map<Object, Object> clones, Object o, Object fieldObject, Field field) throws IllegalAccessException {
-		for (ICloningStrategy strategy : cloningStrategies) {
-			ICloningStrategy.Strategy s = strategy.strategyFor(o, field);
-			if (s == ICloningStrategy.Strategy.NULL_INSTEAD_OF_CLONE) return null;
-			if (s == ICloningStrategy.Strategy.SAME_INSTANCE_INSTEAD_OF_CLONE) return fieldObject;
+	private class FastClonerCloner implements IDeepCloner {
+		private IFastCloner fastCloner;
+		private IDeepCloner cloneInternal;
+
+		FastClonerCloner(IFastCloner fastCloner) {
+			this.fastCloner = fastCloner;
+			this.cloneInternal = deepCloner;
+		}
+
+		public <T> T deepClone(T o, Map<Object, Object> clones) {
+			T clone = (T) fastCloner.clone(o, cloneInternal, clones);
+			if (clones != null) clones.put(o, clone);
+			return clone;
 		}
-		return cloneInternal(fieldObject, clones);
 	}
 
-	@SuppressWarnings("unchecked")
-	private <T> T cloneArray(T o, Map<Object, Object> clones) throws IllegalAccessException {
-		final Class<T> clz = (Class<T>) o.getClass();
-		final int length = Array.getLength(o);
-		final T newInstance = (T) Array.newInstance(clz.getComponentType(), length);
-		if (clones != null) {
-			clones.put(o, newInstance);
+	private static IDeepCloner IGNORE_CLONER = new IgnoreClassCloner();
+	private static IDeepCloner NULL_CLONER = new NullClassCloner();
+
+	private static class IgnoreClassCloner implements IDeepCloner {
+		public <T> T deepClone(T o, Map<Object, Object> clones) {
+			throw new CloningException("Don't call this directly");
```

Comment:
maybe this could return "o" instead of checking 

	if (cloner == IGNORE_CLONER) { ...

Reply:
It could but it is slower as it makes the call site more megamorphic and introduces another method call that is hopefully inlined but seemed to do slightly better. Either way is fine with me.